src/layer_env.rs

```diff
@@ -0,0 +1,882 @@
+    /// effects in buildpack code.
+    ///
+    /// Example:
+    /// ```
+    /// use libcnb::layer_env::{LayerEnv, TargetLifecycle, ModificationBehavior};
+    /// use tempfile::tempdir;
+    /// use std::fs;
+    ///
+    /// let mut layer_env = LayerEnv::new();
+    /// layer_env.insert(TargetLifecycle::Build, ModificationBehavior::Default, "FOO", "bar");
+    /// layer_env.insert(TargetLifecycle::All, ModificationBehavior::Append, "PATH", "some-path");
+    ///
+    /// let mut temp_dir = tempdir().unwrap();
+    /// layer_env.write_to_layer_dir(&temp_dir).unwrap();
+    ///
+    /// assert_eq!(fs::read_to_string(temp_dir.path().join("env.build").join("FOO.default")).unwrap(), "bar");
+    /// assert_eq!(fs::read_to_string(temp_dir.path().join("env").join("PATH.append")).unwrap(), "some-path");
+    /// ```
+    pub fn write_to_layer_dir(&self, layer_dir: impl AsRef<Path>) -> std::io::Result<()> {
+        self.all.write_to_env_dir(layer_dir.as_ref().join("env"))?;
+
+        self.build
+            .write_to_env_dir(layer_dir.as_ref().join("env.build"))?;
+
+        let launch_env_dir = layer_dir.as_ref().join("env.launch");
+
+        self.launch.write_to_env_dir(&launch_env_dir)?;
+
+        for (process_name, delta) in &self.process {
+            delta.write_to_env_dir(launch_env_dir.join(process_name))?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Default for LayerEnv {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Environment variable modification behavior.
+/// ([CNB spec: Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules))
+#[derive(Eq, PartialEq, Debug)]
+pub enum ModificationBehavior {
+    Append,
+    Default,
+    Delimiter,
+    Override,
+    Prepend,
+}
+
+impl Ord for ModificationBehavior {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Explicit mapping used over macro based approach to avoid tying source order of elements
+        // to ordering logic.
+        fn index(value: &ModificationBehavior) -> i32 {
+            match value {
+                ModificationBehavior::Append => 0,
+                ModificationBehavior::Default => 1,
+                ModificationBehavior::Delimiter => 2,
+                ModificationBehavior::Override => 3,
+                ModificationBehavior::Prepend => 4,
+            }
+        }
+
+        index(self).cmp(&index(other))
+    }
+}
+
+impl PartialOrd for ModificationBehavior {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Target lifecycle for an environment variable modification.
+#[derive(Eq, PartialEq, Debug)]
+pub enum TargetLifecycle {
+    All,
+    Build,
+    Launch,
+    Process(String),
+}
+
+#[derive(Eq, PartialEq, Debug)]
+struct LayerEnvDelta {
+    entries: BTreeMap<(ModificationBehavior, OsString), OsString>,
+}
+
+impl LayerEnvDelta {
```

Comment:
Comments and doc tests stopped here 👀

Reply:
Yeah, I only added those for the public API, `LayerEnvDelta` is the first type that is not exposed from this module.